[ARPspoofing] iP

muhammadreyaaz opened the pull request:
# Duke 

## A chatbot to store items from user

* Lightweight project 
* Saves to hard disk 
* Reads from hard disk 

## Developed in an iterative approach every week 

1. In the first week, Duke was designed to store simple input from the user's machine only during the session
2. In the second week, Duke was designed to store more complicated times and tasks
3. In the third week, Duke was given a football fan personality :athletic_shoe: :soccer:
4. Ronaldo **SUI** :goat:

Duke is executable with various levels. Replace the method name with the level name to see previous level iterations

```java 
public static void main(String[] args) {
        Duke duke = new Duke();
        duke.moreOop();
    }
```
- [x] Week 1 
- [x] Week 2 
- [x] Week 3
- [ ] Week 4

Using Git RCS 

```
git status 
git add 
git commit 
````
Duke was built referring to [CS2103/T page](https://nus-cs2103-ay2223s2.github.io/website)

Commit:
Add Gradle support

muhammadreyaaz committed:
Add Printable class which is called in Duke class

Duke reads a list of commands and checks if it is a valid user input.

The commands are fixed and have the same common behaviour.

The common behaviour can cause code duplication and untidiness once Duke
gets more complex.

Extracting the common behaviour into a Printable class allows
abstraction.

Let's store common behaviours into a Printable class so that the Duke
will only need to call Printable. Printable will handle the output.

muhammadreyaaz committed:
Add package and file duke

muhammadreyaaz committed:
Remove files not in package

muhammadreyaaz committed:
Add Tasks and Tasks classes with immutable list

Duke stores the user inputs in an ArrayList.

Duke class will be too cluttered and not abstracted if Tasks and Task
classes are not created.

Let's make the Tasks class as an immutable list with a generic type Task

Using composition which is has-a relationship is preferable over
inheritance

muhammadreyaaz committed:
Add mark and unmark functionality in Task class

Task class has no mark or unmark functionality

This causes the input mark or unmark to be stored instead of the desired
outcome

Implementing mark and undone as attributes for Task allows the user to
mark and unmark a Task

Let's implement attributes mark and unmark and show the status in the
toString function

muhammadreyaaz committed:
Add tracker classes - todos, deadline and events

There are no tracker classes which means if a user keys in todos,
deadline or events, it get recorded as an entry

All three classes have common behaviours

The common behaviours across the three classes can cause code
duplication

Extracting the common behaviour with a superclass as Task prevents
code duplication

Let's make Todos, Events and Deadline inherit from Task and override the
toString method

Using inheritance is preferable over composition in this situation
because the common behaviours are not composable and the superclass
Task is already implemented in previous iterations

muhammadreyaaz committed:
add custom classpath in run script and text output

The default runscript does not point to the correct file path. There is
also a bug which stores multiple lines in one sentence as various keys.

This causes the run script to fail and the bug causes more items in the
list

Modifying the class path and bin path in the run script will allow
compilation and succesful running. No compilation error with regards to
packages

Let's make the class path to run the main function in Duke point to bin
and the package has the remaining path

muhammadreyaaz committed:
Add custom exception handling class & abstraction

There is no exception handling which means the user will not understand
why the input in invalid

Extracting the exceptions into custom classes with a superclass
IllegalArgumentException allows the user to better understand the reason
for the invalid input

Let's pull up DukeException and DukeUnknownException as classes with
IllegalArgumentException as a superclass

muhammadreyaaz committed:
Add error handle for unrecognised commands

Duke error terminates the whole program instead of displaying a message

This does not achieve the desired outcome (the program needs to
continuously run until the user types exit)

The error is catched and a statement is being printed out instead

Let's handle the error in a try-catch block in a function for
encapsulation

muhammadreyaaz committed:
Add delete functionality

muhammadreyaaz committed:
Add custom character messages

muhammadreyaaz committed:
Add save class to store list into hardisk

The user will have an empty startup everytime after terminating the
program

The user's list can be stored into a text file

Let's create a Save class and check if the user has the text file in a
specific directory on all OS. If not, create the file and read / write
into the file everytime the user startups and changes the list

muhammadreyaaz committed:
Add reference text file for storage

muhammadreyaaz committed:
Add localDate parser

muhammadreyaaz committed:
Add save to hardisk features

There is a separate branch that allows the user to save input, read and
write into the hardisk.

Merge into master allows master to be the most updated.

Let's merge branch-Level-7 into master with no fast-forward for an extra
commit.

Merge branch 'branch-Level-7'

muhammadreyaaz committed:
Add localDate parser

Local date parser is present on branch-Level-8 but not in the master
branch.

Master branch needs to be the most updated.

Let's merge branch-Level-8 into master with no fast forward for an extra
commit.

Merge branch 'branch-Level-8'

muhammadreyaaz committed:
Add abstracted classes

Functionalities of parsing user argument and dealing with user arguments
is done by Duke class.

Functionalities can be abstracted to other classes to satsify SOLID
principles

Let's create Ui, Parser, Storage and TaskList classes

muhammadreyaaz committed:
Modify dukeList textfile

muhammadreyaaz committed:
Merge new classes into master branch

Merge branch 'branch-Level-8'

muhammadreyaaz committed:
Remove class files

muhammadreyaaz committed:
Fixed storage error

muhammadreyaaz committed:
Modify dukeList text file

muhammadreyaaz committed:
Merge branch 'add-gradle-support'

muhammadreyaaz committed:
Add gradle support

muhammadreyaaz committed:
Add unit test cases

No unit test case leads to difficulty to deduce the correctness of the
classes

Unit test cases ensure the return output are all intended

Let's create JUnit test cases in test folder and build it with gradle to
check if the unit test cases pass or fails

muhammadreyaaz committed:
Add javadoc

muhammadreyaaz committed:
Modify basic Task

muhammadreyaaz committed:
Add find functionality

muhammadreyaaz committed:
Remove comments

muhammadreyaaz committed:
Add javadoc comments

muhammadreyaaz committed:
Merge javadoc branch

muhammadreyaaz committed:
Add coding standard for javadoc no space

muhammadreyaaz committed:
Merge branch 'branch-A-CodingStandard'

muhammadreyaaz committed:
Remove static find

muhammadreyaaz committed:
Modify find comment

muhammadreyaaz committed:
Merge branch 'branch-Level-9'

muhammadreyaaz committed:
Add basic javafx

muhammadreyaaz committed:
Add javafx basic template in launcher

No graphical user interface is present for the application

Graphical user interface enhances the user's experience

Let's add javafx dependencies in build.gradle and import the relevant
modules in Launcher and Duke

muhammadreyaaz committed:
Modify stage dimensions

muhammadreyaaz committed:
Add iteration 1 echo user command javafx

muhammadreyaaz committed:
Modify javafx to return custom response

muhammadreyaaz committed:
Add image resource folder

muhammadreyaaz committed:
Add javafx custom response without images

muhammadreyaaz committed:
Add image for user and bot

No image is present for both user and bot.

User exeperience is reduced with no images.

Let's add resource stream for each image relative to resource folder
when run from gradle.

muhammadreyaaz committed:
Add rounded profile photo

muhammadreyaaz committed:
Add css file

muhammadreyaaz committed:
Add custom bubble for user and duke

muhammadreyaaz committed:
Fix trail end with no whitespace bug

muhammadreyaaz committed:
Fix xml basic template

muhammadreyaaz committed:
Add basic implementation of iteration 4 without code cleanup

muhammadreyaaz committed:
Add insets padding for dialog box

muhammadreyaaz committed:
Fix profile box issue

muhammadreyaaz committed:
Modify detached head 20bd7d8 padding to be the latest

muhammadreyaaz committed:
Merge branch-No-Bubble to master from modified history

muhammadreyaaz committed:
Add background image

muhammadreyaaz committed:
Remove redundant commented code

muhammadreyaaz committed:
Add checkstyle

muhammadreyaaz committed:
Add checkstyle compatability

muhammadreyaaz committed:
Add variable argumentto get multiple tasks and remove multiple tasks

muhammadreyaaz committed:
Add save to javafx

muhammadreyaaz committed:
Add save with new xml file settings

muhammadreyaaz committed:
Fix region box dynamic height bug

hongshenggg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,286 @@
+ */
+class Parser {
+    //Greetings
+    static String INITIAL_GREETING = "Welcome to Manchester United.";
+    static String FINAL_GREETING = "SUIII, Bye";
+    //Command to display for various user input
+    static String MARK_COMMAND = "SUI, I have marked this task from the training room: ";
+    static String UNMARK_COMMAND = "SUI, I have unmarked this task from the training room: ";
+    static String DELETE_COMMAND = "Tasks successfully deleted. SUI.";
+    static String FIND_COMMAND = "Here are matching tasks in your list";
+    //Whitelist commands
+    static String SHOW_TASKS = "list";
+    static String TERMINATE = "exit";
+    static String MARK = "mark";
+    static String UNMARK = "unmark";
+    static String TODO = "todo";
+    static String DEADLINE = "deadline";
+    static String EVENT = "event";
+    static String DELETE = "delete";
+    static String FIND = "find";
+    //List starts numbering from 1 not from 0
+    static int DECREMENT = 1;
+    //Symbols to display the state of the task to the user beside the type of Task
+    static String MARK_SYMBOL = "X";
+    static String TODO_SYMBOL = "T";
+    static String DEADLINE_SYMBOL = "D";
+    static String EVENT_SYMBOL = "E";
+    //Blacklist
+    static List<String> INVALID_COMMANDS = Arrays.asList("blah", "todo", "deadline", "event");
```
Would be good to label these strings as final since they follow the variable naming conventions for constants.

hongshenggg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,401 @@
+
+        while (true) {
+            String description = sc.next();
+            if (description.equals(Printable.SHOW_TASKS)) {
+                this.tasks.listAllTasks();
+            } else if (description.equals(Printable.TERMINATE)) {
+                Printable.exit();
+            } else if (description.equals(Printable.MARK)) {
+                int taskPosition = sc.nextInt() - Printable.DECREMENT;
+                this.tasks = this.tasks.set(taskPosition, this.tasks.get(taskPosition).markAsDone());
+            } else if (description.equals(Printable.UNMARK)) {
+                int taskPosition = sc.nextInt() - Printable.DECREMENT;
+                this.tasks = this.tasks.set(taskPosition, this.tasks.get(taskPosition).markAsUndone());
+            } else if (description.equals(Printable.TODO)) {
+                description = sc.nextLine();
+                Task newTask = new Todos(description);
+                this.tasks = tasks.add(newTask);
+            } else if (description.equals(Printable.DEADLINE)) {
+                description = sc.nextLine();
+                String[] dateRange = description.split("/");
+                Task newTask = new Deadline(dateRange[0], dateRange[1].replaceFirst("by", ""));
+                this.tasks = tasks.add(newTask);
+            } else if (description.equals(Printable.EVENT)) {
+                description = sc.nextLine();
+                String[] dateRange = description.split("/");
+                Task newTask = new Events(dateRange[0], dateRange[1].replaceFirst("from", ""),
+                        dateRange[2].replaceFirst("to", ""));
+                this.tasks = tasks.add(newTask);
+            } else {
+                Task newTask = new Task(description + sc.nextLine());
+                this.tasks = tasks.add(newTask);
+            }
```
Perhaps switch-case statements might look better here?

hongshenggg commented on src/main/java/duke/Save.java:
```diff
@@ -0,0 +1,111 @@
+import java.util.Scanner;
+import java.io.File;
+
+class Save {
```
Some header comments in this class for documentation would help

hongshenggg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,401 @@
+import javafx.fxml.FXML;
+
+/**
+ * Duke is the main class that directly handles the user input, and abstract
```
Good job on writing well descriptive documentation

hongshenggg reviewed (commented):
LGTM. Just some nits to fix.

muhammadreyaaz committed:
Modify dukeList text

muhammadreyaaz committed:
Modify dukeList

muhammadreyaaz committed:
Modify dukelist

muhammadreyaaz committed:
Modify dukelist

muhammadreyaaz committed:
Modify dukelist merge with branch level 10

muhammadreyaaz committed:
Merge branch-Level-10 to fix region box height calculated dimension

muhammadreyaaz committed:
Remove magic string and numbers from Storage

muhammadreyaaz committed:
Modify gitignore to ignore dukeList text file

muhammadreyaaz committed:
Modify gitignore file

muhammadreyaaz committed:
Modify gitignore

muhammadreyaaz committed:
Remove all usued methods in Duke

Methods from previous iterations exist, all the way from week 1

Duke file is very long and messy

Remove all unused methods, and retrieve it from previous commits if
required

muhammadreyaaz committed:
Remove Printable, Save and Tasks classes

Printable, Save and Tasks class still exists

It is no longer in use, making the file cluttered

Let's delete the files, and retrieve them from previous commits if
required

muhammadreyaaz committed:
Remove variable depending on Tasks class

muhammadreyaaz committed:
Add code quality for background image of dialogbox

Background image is hardcoded from url in fxml file

Url is long and makes the fxml file messy

Let's save the image in resources/images and use the dynamic filepath to
create a background image

muhammadreyaaz committed:
Add basic template for ci pipeline

muhammadreyaaz committed:
Add task scheduler with mock timing data

No reminder functionality

User will thus have to manually sieve through which is the upcoming task

Let's use ExecutorService and PriorityBlockingQueue to remind the user
of the next upcoming event

muhammadreyaaz committed:
Provide minor styling for labels

muhammadreyaaz committed:
Fix bug where custom bot command not shown on Gui

muhammadreyaaz committed:
Add multipage navigation

Single page application

User will be unaware of what the application does

Let's add a new scene that explains the app and a button to navigate to
the chat application

muhammadreyaaz committed:
Add multipage

muhammadreyaaz committed:
Add welcome page

No welcome page.

User experience is reduced.

Let's add a login page with username and password.

muhammadreyaaz committed:
Provide main method in duke to pass ci

muhammadreyaaz committed:
Modify runtest script

muhammadreyaaz committed:
Modify runtest script

muhammadreyaaz committed:
Modify runtest script

muhammadreyaaz committed:
Modify build script

muhammadreyaaz committed:
Provide correct actual and expected text files

Runtest script comparing wrong version of expected file.

A correct expected text file is necessary for ci to be successful.

Let's,
* add the correct actual txt file
* remove old expected txt file

muhammadreyaaz committed:
Modify runtest bat and Remove redundant code

muhammadreyaaz committed:
Merge pull request #1 from ARPspoofing/branch-ci

Modify runtest bat and Remove redundant code

lennoxtr commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+ */
+
+class Deadline extends Task {
+    protected String by;
```
I think it would be better if you use a more intuitive name. In addition, is there any reason why you store the deadline as String instead of LocalDate, and then convert it to LocalDate later with the method localDateParser()?
```suggestion
    protected String endDate;
```

lennoxtr commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,53 @@
+        super(description, isDone);
+        this.by = by;
+    }
+    String localDateParser() {
```
A javadoc explaining the method purpose would be great here

lennoxtr commented on src/main/java/duke/Events.java:
```diff
@@ -0,0 +1,104 @@
+ *  @see Task
+ *  @since 11
+ */
+class Events extends Task {
```
I suggest changing the class name to Event (singular)
```suggestion
class Event extends Task {
```

lennoxtr commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,292 @@
+     * @return TaskList New list of tasks with the latest being the todo task added
+     * @exception DukeException
+     */
+    static TaskList<Task> toDo(Scanner sc, TaskList<Task> tasks) {
```
perhaps addToDo would be a better name here
```suggestion
    static TaskList<Task> addToDo(Scanner sc, TaskList<Task> tasks) {
```

lennoxtr commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,292 @@
+     * @return TaskList New list of tasks with the latest being the todo task added
+     * @exception DukeException
+     */
+    static TaskList<Task> deadline(String description, String date, TaskList<Task> tasks) {
```
A more intuitive name would be better here too
```suggestion
    static TaskList<Task> addDeadline(String description, String date, TaskList<Task> tasks) {
```

lennoxtr commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,292 @@
+     * @return TaskList New list of tasks with the latest being the todo task added
+     * @exception DukeException
+     */
+    static TaskList<Task> events(String description, String from, String to, TaskList<Task> tasks) {
```
More intuitive name here would be nice too! 
```suggestion
    static TaskList<Task> addEvent(String description, String from, String to, TaskList<Task> tasks) {
```

lennoxtr commented on src/main/java/duke/Recur.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.prefs.*;
+import java.util.regex.*;
+ import java.util.stream.*;
+
```
a javadoc explaining the purpose of this class would be helpful

lennoxtr commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,199 @@
+     * Getter to get the current set of tasks
+     * @return TaskList<Task>
+     */
+    TaskList<Task> getTasks() {
```
very intuitive naming of methods here. Keep up the good work!

lennoxtr commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,167 @@
+     * @param newTask Task to add
+     * @return TaskList<T>
+     */
+    TaskList<T> add(Task newTask) {
```
perhaps addTask would be a better name here
```suggestion
    TaskList<T> addTask(Task newTask) {
```

lennoxtr commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,167 @@
+     * Get a task in a specific position
+     * @return positionOfTask
+     */
+    Task get(int index) {
```
getTask would be a better name  here
```suggestion
    Task getTask(int index) {
```

lennoxtr commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,167 @@
+     *
+     * @return TaskList<T>
+     */
+    TaskList<T> set(int index, Task task) {
```
setTaskStatus would be a more appropriate name in my opinion
```suggestion
    TaskList<T> setTaskStatus(int index, Task task) {
```

lennoxtr commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,35 @@
+package duke;
+
+import org.junit.jupiter.api.Test;
+
```
should be imports be grouped together?

lennoxtr commented on src/test/java/duke/EventsTest.java:
```diff
@@ -0,0 +1,45 @@
+package duke;
+
+import org.junit.jupiter.api.Test;
+
```
should the imports be grouped together?

lennoxtr reviewed (commented):
Keep up the good work! Just need a few changes regarding class names, and you're good to go!

muhammadreyaaz committed:
Add recur priority queue

muhammadreyaaz committed:
Add recur task as a subclass of events

There is no recur task

The user will always have to key in recurring task

Let's:
    * Add ExecutorService for multi threading
    * Add PriorityBlockedQueue
    * Add new Recur class

muhammadreyaaz committed:
Fix recurring bug, trailing white space bug and add assert

muhammadreyaaz committed:
Add recur event to GUI

Recur event is only displayed on terminal on a separate thread

The GUI only uses a specific thread.

Let's:
    * Remove using executorService for recur events
    * Use timeline instead on the dialog container
    * Comment out the executor service in Task Scheduler
    * Add Timeline in Task Scheduler

muhammadreyaaz committed:
Remove redudant code

muhammadreyaaz committed:
Recur bug fixed

muhammadreyaaz committed:
Add recur touchup and remove redundant code

muhammadreyaaz committed:
Remove redundant code fromparser

muhammadreyaaz committed:
Remove unnecessary code from Duke

muhammadreyaaz committed:
Remove unnecessary code from Duke and InitialWindow

muhammadreyaaz committed:
Remove unnecessary code from Main

muhammadreyaaz committed:
Remove unnecessary code from Ui

muhammadreyaaz committed:
Remove unnecessary code from Ui

muhammadreyaaz committed:
Provide abstraction methods for get response

All the functionality is in the same method

Reviewers might find it difficult to understand

Let's provide method abstraction with method names stating the
abstracted process

muhammadreyaaz committed:
Provide abstraction for styling

Styling is in the constructor

Abstracting the styling is better for redeability

Let's move the styling into another method, and call it in the
constructor

muhammadreyaaz committed:
Remove redundant code

muhammadreyaaz committed:
Fix recur bug

muhammadreyaaz committed:
Provide JUnit test cases

muhammadreyaaz committed:
Provide task scheduler JUnit test

muhammadreyaaz committed:
Provide task scheduler test

muhammadreyaaz committed:
Provide ui junit test cases

muhammadreyaaz committed:
Provide test cases for Storage

muhammadreyaaz committed:
Provide test cases for Parser class

muhammadreyaaz committed:
Provide test cases for Parser class

muhammadreyaaz committed:
Provide JUnit for DialogBox

muhammadreyaaz committed:
Provide test cases for Main

muhammadreyaaz committed:
Remove unnecessary commented out code in test case files

muhammadreyaaz committed:
Add javadoc for deadline class

muhammadreyaaz committed:
Add javadoc comments for dialogbox class

muhammadreyaaz committed:
Add javadoc comments for duke class

muhammadreyaaz committed:
Provide javadoc comments for event class

muhammadreyaaz committed:
Provide javadoc comments for initial window class

muhammadreyaaz committed:
Provide javadoc for launcher and mainwindow

muhammadreyaaz committed:
Provide javadoc for recur class

muhammadreyaaz committed:
Provide javadoc comments for Storage class

muhammadreyaaz committed:
Remove executorservice from taskscheduler. Retrieve from here if needed again

muhammadreyaaz committed:
Provide javadoc comments for task scheduler

muhammadreyaaz committed:
Remove unnecessary things from ui class

muhammadreyaaz committed:
Provide javadoc comments for todo class

muhammadreyaaz committed:
Provide assymetric styling for user dialog background

muhammadreyaaz committed:
Add png to docs folder

muhammadreyaaz committed:
Provide readme

muhammadreyaaz committed:
Provide comprehensivereadme in user guide format

muhammadreyaaz committed:
Provide comprehensive readme

muhammadreyaaz committed:
Modify image path in user guide

muhammadreyaaz committed:
Remove verbose in readme main page

muhammadreyaaz committed:
Remove verbose in readme main page

muhammadreyaaz committed:
Remove yaml error in user guide

muhammadreyaaz committed:
Modify all files to abide by checkstyle

muhammadreyaaz committed:
Remove sudo from brew install dos2unix

muhammadreyaaz committed:
Make ci pipeline pass

muhammadreyaaz committed:
Make ci pipeline pass

muhammadreyaaz committed:
Make ci pipeline pass

muhammadreyaaz committed:
Make ci pipeline pass

muhammadreyaaz committed:
Make ci pipeline pass

muhammadreyaaz committed:
Update readme to tell user that any username and password is valid
